File: r-mos-frontend/src/api/agent.ts
// Agent API Client
// P0: Frontend integration for Agent services

import client from './client';

// Types matching backend services

export interface AgentRequest {
  request_id?: string;
  user_id: string;
  message: string;
  context?: Record<string, unknown>;
}

export interface AgentResponse {
  response_id: string;
  request_id: string;
  message: string;
  action_suggested?: Record<string, unknown>;
  confidence: number;
  evidence_refs: string[];
}

const isObjectRecord = (value: unknown): value is Record<string, unknown> =>
  typeof value === 'object' && value !== null;

const readString = (value: unknown, fallback: string): string =>
  typeof value === 'string' ? value : fallback;

const readNumber = (value: unknown, fallback: number): number =>
  typeof value === 'number' ? value : fallback;

const readStringArray = (value: unknown): string[] =>
  Array.isArray(value) ? value.filter((item): item is string => typeof item === 'string') : [];

export interface NextAction {
  action_id: string;
  action_type: string;
  target?: string;
  explanation: string;
  risk_level: string;
  evidence_required: string[];
}

export interface CoachOutput {
  next_action?: NextAction;
  explanation: string;
  risk_events: Record<string, unknown>[];
  confidence: number;
}

export interface DiagnoserOutput {
  root_cause?: string;
  root_cause_confidence: number;
  evidence_refs: string[];
  intervention?: Record<string, unknown>;
  baseline_comparison?: Record<string, unknown>;
}

export interface KnowledgeEntry {
  id: string;
  type: string;
  status: string;
  title: string;
  content: string;
  scope?: {
    device_model: string[];
    part_type: string[];
    version_range?: string;
    scenario: string[];
  };
  contraindications?: {
    device_model: string[];
    part_material: string[];
    conditions: string[];
  };
  risk_level: string;
  confidence?: {
    evidence_count: number;
    success_rate: number;
    reviewer_count: number;
  };
  created_at: number;
  created_by: string;
}

export interface KnowledgeUploadJob {
  job_id: string
  status: string
  filename?: string
  content_type?: string
  size_bytes?: number
  brand?: string | null
}

// Agent Orchestrator API

export const sendAgentRequest = async (request: AgentRequest): Promise<AgentResponse> => {
  // P2-1: Use unified /agent/execute endpoint (message mode)
  const response = await client.post<AgentExecuteResponse>('/agent/execute', {
    user_id: request.user_id,
    mode: 'message',
    message: request.message,
    context: request.context,
  });

  // Convert to legacy response format for backward compatibility
  const data = response.data;
  const result = isObjectRecord(data.result) ? data.result : {};
  const messageFromResult = readString(result.message, readString(result.response, JSON.stringify(result)));
  const actionSuggested = isObjectRecord(result.action) ? result.action : undefined;
  return {
    response_id: data.trace_id,
    request_id: request.request_id || data.trace_id,
    message: messageFromResult,
    action_suggested: actionSuggested,
    confidence: readNumber(result.confidence, 0.9),
    evidence_refs: readStringArray(result.evidence_refs),
  };
};

export const getTaskStatus = async (userId: string): Promise<Record<string, unknown>> => {
  const response = await client.get<Record<string, unknown>>(`/agent/task-status/${userId}`);
  return response.data;
};

// Coach Agent API

export const getCoachRecommendation = async (
  taskId: string,
  currentStep: number,
  stepHistory: Record<string, unknown>[],
  traineeAction?: Record<string, unknown>
): Promise<CoachOutput> => {
  const response = await client.post<CoachOutput>('/agent/coach/recommend', {
    task_id: taskId,
    current_step: currentStep,
    step_history: stepHistory,
    trainee_action: traineeAction,
  });
  return response.data;
};

// Diagnoser Agent API

export const diagnoseError = async (
  taskId: string,
  errorHistory: Record<string, unknown>[],
  actionHistory: Record<string, unknown>[],
  evidenceRefs: string[]
): Promise<DiagnoserOutput> => {
  const response = await client.post<DiagnoserOutput>('/agent/diagnoser/diagnose', {
    task_id: taskId,
    error_history: errorHistory,
    action_history: actionHistory,
    evidence_refs: evidenceRefs,
  });
  return response.data;
};

// Knowledge Governance API

export const searchKnowledge = async (query: {
  query?: string;
  device_model?: string;
  part_type?: string;
  status?: string;
}): Promise<{ results: KnowledgeEntry[] }> => {
  const response = await client.post<{ results: KnowledgeEntry[] }>('/agent/knowledge/search', query);
  return response.data;
};

export const createKnowledge = async (data: {
  title: string;
  content: string;
  type: string;
  scope?: Record<string, unknown>;
  risk_level?: string;
}): Promise<KnowledgeEntry> => {
  const response = await client.post<KnowledgeEntry>('/agent/knowledge', data);
  return response.data;
};

export const submitKnowledgeForReview = async (entryId: string): Promise<void> => {
  await client.post(`/agent/knowledge/${entryId}/submit`);
};

export const approveKnowledge = async (
  entryId: string,
  decision: 'approve' | 'reject',
  feedback?: string
): Promise<void> => {
  await client.post(`/agent/knowledge/${entryId}/approve`, {
    decision,
    feedback,
  });
};

export const uploadKnowledgeFile = async (file: File, brand?: string): Promise<KnowledgeUploadJob> => {
  const formData = new FormData()
  formData.append('file', file)
  if (brand) {
    formData.append('brand', brand)
  }
  const response = await client.post<KnowledgeUploadJob>('/agent/knowledge/upload', formData, {
    headers: {
      'Content-Type': 'multipart/form-data',
    },
  })
  return response.data
}

export const getKnowledgeUploadJob = async (jobId: string): Promise<KnowledgeUploadJob> => {
  const response = await client.get<KnowledgeUploadJob>(`/agent/knowledge/upload/${jobId}`)
  return response.data
}

// Multi-Agent Coordination API

export const coordinateAgents = async (
  taskId: string,
  userId: string,
  action: string,
  context: Record<string, unknown>
): Promise<Record<string, unknown>> => {
  const response = await client.post<Record<string, unknown>>('/agent/coordinate', {
    task_id: taskId,
    user_id: userId,
    action,
    context,
  });
  return response.data;
};

// Evidence Enforcement API

export const getEvidenceStatus = async (stepId: string): Promise<Record<string, unknown>> => {
  const response = await client.get<Record<string, unknown>>(`/agent/evidence/status/${stepId}`);
  return response.data;
};

export const collectEvidence = async (
  stepId: string,
  evidenceId: string,
  evidenceType: string
): Promise<void> => {
  await client.post('/agent/evidence/collect', {
    step_id: stepId,
    evidence_id: evidenceId,
    evidence_type: evidenceType,
  });
};

export const canProceedToNextStep = async (stepId: string): Promise<{ allowed: boolean; reason: string }> => {
  const response = await client.get<{ allowed: boolean; reason: string }>(`/agent/evidence/can-proceed/${stepId}`);
  return response.data;
};

// ============ P2-1: Unified Agent Execute API ============

export type AgentExecuteMode = 'command' | 'message' | 'auto';

export interface AgentExecuteRequest {
  // Common
  user_id: string;
  mode?: AgentExecuteMode;

  // Command mode fields
  intent?: string;
  tool_name?: string;
  tool_args?: Record<string, unknown>;
  skill_id?: string;
  side_effects?: string[];
  input_text?: string;

  // Message mode fields
  message?: string;

  // Shared fields
  context?: Record<string, unknown>;
  resource_ref?: Record<string, unknown>;
  policy_context?: Record<string, unknown>;
  intent_classification?: string;
  trace_id?: string;
  idempotency_key?: string;
}

export interface AgentExecuteResponse {
  status: 'success' | 'pending_approval' | 'error';
  result?: Record<string, unknown>;
  trace_id: string;
  from_cache: boolean;
  approval_id?: number;
  mode_used: 'command' | 'message';
}

/**
 * Unified Agent Execute API - P2-1 Convergence
 *
 * Replaces both:
 * - POST /ai/commands (Command mode)
 * - POST /agent/v2/request (Message mode)
 *
 * Auto-detects mode if not specified.
 */
export const agentExecute = async (request: AgentExecuteRequest): Promise<AgentExecuteResponse> => {
  const response = await client.post<AgentExecuteResponse>('/agent/execute', request);
  return response.data;
};

File: r-mos-backend/app/api/v1/endpoints/agent.py
"""
Agent API Endpoints
P0: Frontend integration for Agent services
"""

from fastapi import APIRouter, Depends, Request
from typing import Dict, Any
import uuid
import logging

from app.core.database import get_db
from sqlalchemy.ext.asyncio import AsyncSession
from app.models.command_runtime import Command, AIToolCall
from app.models.approval import Approval
from app.services.tool_executor import validate_tool_request_security
from app.services.access_control import log_allow_event, log_deny_event
from app.services.authz_guard import ActorContext, get_current_actor
from app.api.v1.endpoints.ai_commands import _plan_tool_call

logger = logging.getLogger(__name__)

from app.services.agent_service import orchestrator
from app.services.coach_agent import coach_agent
from app.services.diagnoser_agent import diagnoser_agent
from app.services.multi_agent_coordinator import multi_agent_coordinator
from app.services.orchestrator_v2 import orchestrator_v2
from app.services.authz_guard import require_permission
from app.schemas.agent import (
    CoachRecommendRequest,
    DiagnoseRequest,
    CoordinateRequest,
    AgentExecuteMode,
    AgentExecuteRequest,
    AgentExecuteResponse,
)
from app.api.v1.endpoints import agent_knowledge, agent_evidence, agent_v2, agent_governance
from app.api.v1.endpoints.agent_v2 import _require_agent_permission

# Re-export symbols that tests access via `agent_endpoints.<name>`
from app.services.knowledge_governance import knowledge_governance  # noqa: F401
from app.api.v1.endpoints.agent_knowledge import knowledge_upload_jobs  # noqa: F401

router = APIRouter(prefix="/agent", tags=["agent"])

router.include_router(agent_knowledge.router)
router.include_router(agent_evidence.router)
router.include_router(agent_v2.router)
router.include_router(agent_governance.router)


@router.get("/task-status/{user_id}")
async def get_task_status(
    user_id: str,
    _: None = Depends(require_permission("agent:read", required_role="agent_user")),
):
    """Get current task status for user"""
    status = orchestrator.get_task_status(user_id)
    return status


# ============ Coach Agent Endpoints ============

@router.post("/coach/recommend", response_model=Dict[str, Any])
async def get_coach_recommendation(request: CoachRecommendRequest, _: None = Depends(require_permission("agent:execute"))):
    """Get coach agent recommendation for next action"""
    output = coach_agent.analyze_and_recommend(
        task_id=request.task_id,
        current_step=request.current_step,
        step_history=request.step_history,
        trainee_action=request.trainee_action
    )

    return {
        "next_action": output.next_action.model_dump() if output.next_action else None,
        "explanation": output.explanation,
        "risk_events": output.risk_events,
        "confidence": output.confidence,
        "reasoning": output.reasoning,
    }


# ============ Diagnoser Agent Endpoints ============

@router.post("/diagnoser/diagnose", response_model=Dict[str, Any])
async def diagnose_error(request: DiagnoseRequest, _: None = Depends(require_permission("agent:execute"))):
    """Diagnose root cause of errors"""
    output = diagnoser_agent.diagnose(
        task_id=request.task_id,
        error_history=request.error_history,
        action_history=request.action_history,
        available_evidence=request.evidence_refs
    )

    return {
        "root_cause": output.root_cause.cause_type.value if output.root_cause else None,
        "root_cause_confidence": output.root_cause.confidence if output.root_cause else 0,
        "evidence_refs": output.evidence_refs,
        "intervention": output.intervention,
        "baseline_comparison": output.baseline_comparison,
        "confidence": output.confidence,
        "reasoning": output.reasoning,
    }


# ============ Multi-Agent Coordination Endpoints ============

@router.post("/coordinate")
async def coordinate_agents(request: CoordinateRequest, _: None = Depends(require_permission("agent:execute"))):
    """Coordinate multiple agents"""
    result = await multi_agent_coordinator.coordinate(
        task_id=request.task_id,
        user_id=request.user_id,
        action=request.action,
        context=request.context
    )

    return {
        "task_id": result.task_id,
        "user_id": result.user_id,
        "final_action": result.final_action,
        "consensus": result.consensus,
        "conflicts": result.conflicts,
        "execution_time_ms": result.execution_time_ms,
    }


# ============ P2-1: Unified Agent Execute Schema ============

def _detect_mode(request: AgentExecuteRequest) -> str:
    """Auto-detect execution mode based on request fields"""
    if request.mode != AgentExecuteMode.AUTO:
        return request.mode

    # Auto-detect: command mode takes priority if command fields are provided
    if request.tool_name or request.intent:
        return AgentExecuteMode.COMMAND
    if request.message:
        return AgentExecuteMode.MESSAGE

    # Default to message mode
    return AgentExecuteMode.MESSAGE


@router.post("/execute", response_model=AgentExecuteResponse, deprecated=False)
async def execute_agent(
    request: AgentExecuteRequest,
    http_request: Request,
    db: AsyncSession = Depends(get_db),
    actor: ActorContext = Depends(get_current_actor),
):
    """Unified Agent Execute Endpoint - P2-1 Convergence

    This endpoint handles both command execution and message processing.
    """
    # Generate trace_id if not provided
    trace_id = request.trace_id or str(uuid.uuid4())[:8]
    http_request.state.trace_id = trace_id

    # Detect execution mode
    mode = _detect_mode(request)
    required_permission = "agent:execute" if mode == AgentExecuteMode.COMMAND else "agent:read"
    await _require_agent_permission(
        http_request,
        db,
        actor,
        permission_key=required_permission,
    )
    logger.info(f"[P2-1] Agent execute: mode={mode}, user_id={request.user_id}, trace_id={trace_id}")

    if mode == AgentExecuteMode.COMMAND:
        # A-mode: Handle as command execution
        try:
            # Build CommandCreateRequest-like payload
            command_payload = {
                "intent": request.intent or "dispatch",
                "input_text": request.input_text,
                "skill_id": request.skill_id,
                "tool_name": request.tool_name,
                "tool_args": request.tool_args,
                "side_effects": request.side_effects,
            }

            # Create Command record
            # 注意：Command 模型的用户字段是 actor_user_id（非 user_id），且 trace_id 必填、
            # 无 input_text 列。此前误用 user_id/input_text 会在构造时抛 TypeError 被吞成 error。
            command = Command(
                trace_id=trace_id,
                actor_user_id=request.user_id,
                intent=command_payload["intent"],
                skill_id=command_payload.get("skill_id"),
                status="pending",
            )
            db.add(command)
            await db.commit()
            await db.refresh(command)

            # Plan tool call
            from app.api.v1.endpoints.ai_commands import CommandCreateRequest

            cmd_req = CommandCreateRequest(**command_payload)
            planned_tool = _plan_tool_call(cmd_req)

            # Validate security
            try:
                validate_tool_request_security(
                    tool_name=planned_tool.tool_name,
                    tool_args=planned_tool.tool_args,
                )
            except Exception as sec_error:
                await log_deny_event(
                    db,
                    http_request,
                    action="tool_call_failed",
                    resource_type="Skill",
                    resource_id=planned_tool.skill_id or planned_tool.tool_name,
                    reason=str(sec_error),
                    actor_user_id=str(actor.user_id),
                    skill_id=planned_tool.skill_id,
                    tool_call_args=planned_tool.tool_args,
                    side_effects_applied=list(planned_tool.side_effects),
                )
                raise

            # Log allow event
            await log_allow_event(
                db,
                http_request,
                action="command_created",
                actor_user_id=str(actor.user_id),
                resource_type="Command",
                resource_id=command.id,
                reason="unified_agent_execute",
            )

            # Create tool call
            tool_call = AIToolCall(
                command_id=command.id,
                trace_id=trace_id,
                actor_user_id=str(actor.user_id),
                skill_id=planned_tool.skill_id,
                tool_name=planned_tool.tool_name,
                side_effects=list(planned_tool.side_effects),
                status="pending",
                approval_id=None,
            )
            db.add(tool_call)
            await db.commit()
            await db.refresh(tool_call)

            # Handle side effects - require approval
            approval_id = None
            result_status = "success"
            result_data = {
                "command_id": command.id,
                "tool_call_id": tool_call.id,
                "tool_name": planned_tool.tool_name,
                "skill_id": planned_tool.skill_id,
                "status": "pending",
            }

            if planned_tool.side_effects:
                approval = Approval(
                    trace_id=trace_id,
                    command_id=command.id,
                    tool_call_id=tool_call.id,
                    status="pending",
                    reason="awaiting_approval",
                    created_by_user_id=str(actor.user_id),
                )
                db.add(approval)
                await db.commit()
                await db.refresh(approval)

                command.approval_id = approval.id
                tool_call.approval_id = approval.id
                command.status = "waiting_approval" if planned_tool.via_planner else "pending_approval"
                await db.commit()

                approval_id = approval.id
                result_status = "pending_approval"
                result_data["status"] = "waiting_approval"
                result_data["approval_id"] = approval.id

            return AgentExecuteResponse(
                status=result_status,
                result=result_data,
                trace_id=trace_id,
                from_cache=False,
                approval_id=approval_id,
                mode_used=AgentExecuteMode.COMMAND,
            )

        except Exception as e:
            logger.error(f"[P2-1] Command mode error: {e}")
            return AgentExecuteResponse(
                status="error",
                result={"error": str(e)},
                trace_id=trace_id,
                from_cache=False,
                approval_id=None,
                mode_used=AgentExecuteMode.COMMAND,
            )

    else:
        # B-mode: Handle as message (via OrchestratorV2)
        try:
            response = await orchestrator_v2.process_request(
                user_id=request.user_id,
                message=request.message or "",
                resource_ref=request.resource_ref,
                policy_context=request.policy_context,
                intent_classification=request.intent_classification,
                telemetry_payload=request.telemetry_payload,
                trace_id=trace_id,
                idempotency_key=request.idempotency_key,
            )

            return AgentExecuteResponse(
                status="success",
                result=response,
                trace_id=trace_id,
                from_cache=response.get("from_cache", False),
                approval_id=response.get("approval_id"),
                mode_used=AgentExecuteMode.MESSAGE,
            )

        except Exception as e:
            logger.error(f"[P2-1] Message mode error: {e}")
            return AgentExecuteResponse(
                status="error",
                result={"error": str(e)},
                trace_id=trace_id,
                from_cache=False,
                approval_id=None,
                mode_used=AgentExecuteMode.MESSAGE,
            )
